server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -219,83 +218,148 @@ public void stop()
-                        getSegmentsTable(),
-                        segment.getId()
+                        "SELECT payload FROM %1$s WHERE dataSource = :dataSource AND id = :id",
+                        getSegmentsTable()
                     )
-                );
-              }
-              batch.execute();
-
-              return null;
-            }
+                )
+                .setFetchSize(connector.getStreamingFetchSize())
+                .bind("dataSource", dataSource)
+                .bind("id", segmentId)
+                .map(ByteArrayMapper.FIRST)
+                .spliterator(), false
+            ).findFirst().orElseThrow(
+                () -> new UnknownSegmentIdException(StringUtils.format("Cannot find segment id [%s]", segmentId))
+            ), DataSegment.class);
           }
-      );
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+        }).collect(Collectors.toList()))
+    );
+  }
+
+  private Stream<SegmentId> segmentIdsForInterval(
+      final VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline,
+      final Interval interval
+  )
+  {
+    return versionedIntervalTimeline.lookup(interval).stream().flatMap(
+        objectHolder -> StreamSupport.stream(objectHolder.getObject().spliterator(), false).map(
+            dataSegmentPartitionChunk -> dataSegmentPartitionChunk.getObject()
+        )
+        .filter(segment -> interval.contains(segment.getInterval()))
+        .map(segment -> segment.getId())
+    );
+  }
+
+  @Override
+  public boolean enableDataSource(final String dataSource)
+  {
+    try {
+      return enableSegments(dataSource, Intervals.ETERNITY) != 0;
     }
     catch (Exception e) {
       log.error(e, "Exception enabling datasource %s", dataSource);
       return false;
     }
+  }
 
-    return true;
+  @Override
+  public int enableSegments(final String dataSource, final Interval interval)
+  {
+    VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline = getVersionedIntervalTimeline(dataSource, interval);
+
+    return enableSegments(
+        segmentIdsForInterval(versionedIntervalTimeline, interval).collect(Collectors.toSet()),
+        versionedIntervalTimeline
+    );
+  }
+
+  @Override
+  public int enableSegments(final String dataSource, final Collection<String> segmentIds)
+  {
+    VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline = getVersionedIntervalTimeline(dataSource, segmentIds);
+    versionedIntervalTimeline.getAllTimelineEntries().keySet().stream().reduce(
+        (acc, val) -> acc.withStartMillis(Math.min(acc.getStartMillis(), val.getStartMillis())).withEndMillis(Math.max(acc.getEndMillis(), val.getEndMillis()))
+    ).ifPresent(interval -> VersionedIntervalTimeline.addSegments(
+        versionedIntervalTimeline,
+        dataSources.get(dataSource).getSegments().stream().filter(segment -> interval.contains(segment.getInterval())).iterator()
```

Comment:
The latter part that adds segments from the datasource should use interval.overlaps to deal with partial overshadowing

Reply:
Yes and as soon as I noticed this I also noticed my fix to do overlap on the SQL query gets non-enabled segments so both Interval and SegmentIds calls need to use this logic.

Reply:
This should all be correct now.